### PR TITLE
test(general): add crash consistency test to validate repository consistency when kopia stops abruptly

### PR DESCRIPTION
**Pull request by MinyiZou**
The main objective is to verify and ensure that the Kopia backup tool does not cause corruption in the repository when a crash occurs, particularly triggered by an unhandled SIGKILL signal. Most important thing is to support a dedicated pipeline for stress and longevity testing in Crash Consistency Testing.

### CODE REVIEW

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 1]
```diff
@@ -274,6 +274,21 @@ func (bm *BlobManipulator) SetUpSystemUnderTest() error {
 	return nil
 }
 
+// GenerateRandomFiles connects or creates a kopia repo, writes random data in source directory,
+// creates snapshots of the source directory.
```
Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:278]:
```suggestion
// GenerateRandomFiles connects or creates a Kopia repository that writes random data in source directory.
// Tests can later create snapshots from the source directory.
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+)
+
+func TestConsistencyWhenKill9AfterModify(t *testing.T) {
+	// create, connect repository
+	dataRepoPath := path.Join(*repoPathPrefix, dirPath+dataPath)
+	baseDir := makeBaseDir()
+	bm, err := blobmanipulator.NewBlobManipulator(baseDir, dataRepoPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping recovery tests because KOPIA_EXE is not set")
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:32]: You want to mark this test as skipped instead of success, so better to use the testing skip interface for that.

```suggestion
			t.Skip("Skipping recovery tests because KOPIA_EXE is not set")
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+		} else {
+			log.Println("Error creating Blob Manipulator:", err)
+		}
+		os.Exit(0)
+	}
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:37]: There is a nice error checking interface from `github.com/stretchr/testify/require` package which you can import and use to make this error handling a bit simpler:

```suggestion
	if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
		t.Skip("Skipping recovery tests because KOPIA_EXE is not set")
	}

	require.NoError(t, err)
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	bm.DataRepoPath = dataRepoPath
+
+	// generate 10M * 10 dataset
+	fileSize := 1 * 1024 * 1024
+	numFiles := 100
+	err = bm.GenerateRandomFiles(fileSize, numFiles)
+	if err != nil {
+		log.Println("Error in creating snapshot data", err)
+		t.FailNow()
+	}
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:48]: Same suggestion as above, I'll stop point these out now. You can use the `require` interface on most other changes as well.

```suggestion
	require.NoError(t, err, "failed creating source data for the snapshot operation")
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	// create snapshot successfully at the first time
+	// populate the kopia repo under test with random snapshots
+	snapID, _, err := bm.TakeSnapshot(bm.PathToTakeSnapshot)
+	if err != nil {
+		log.Println("Error in creating snapshot", err)
+		t.FailNow()
+	}
+	dst := getRootDir(bm.PathToTakeSnapshot)
+
+	preResotrePath := filepath.Join(dirPath, "restore_pre")
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:59]:
```suggestion
	preRestorePath := filepath.Join(dirPath, "restore_pre")
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	fileSize := 1 * 1024 * 1024
+	numFiles := 100
+	err = bm.GenerateRandomFiles(fileSize, numFiles)
+	if err != nil {
+		log.Println("Error in creating snapshot data", err)
+		t.FailNow()
+	}
+
+	// create snapshot successfully at the first time
+	// populate the kopia repo under test with random snapshots
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:51]:
```suggestion
	// Create the first snapshot and expect it to run successfully.
	// This will populate the repository under test with initial data snapshot.
```

File: tests/robustness/crash_consistency_test/crash_consistency_test.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+	}
+	dst := getRootDir(bm.PathToTakeSnapshot)
+
+	preResotrePath := filepath.Join(dirPath, "restore_pre")
+	// try to restore a snapshot named restore_pre
+	stdout, err := bm.RestoreGivenOrRandomSnapshot("", preResotrePath)
+	if err != nil {
+		log.Println("Error restoring the kopia repository:", stdout, err)
+		t.FailNow()
+	}
```
Comment (Shrekster) [inline tests/robustness/crash_consistency_test/crash_consistency_test.go:65]: Hmmm. I am trying to understand this step a little better. What do you think is the purpose of trying to perform a restore before the second snapshot ?

I think you maybe have some valid reason here, I just would like to understand it better.

**Comment by chaitalisg** [comment]
A lot of the code here is similar to the `recovery` test package. Can you move the test to that package and pull out the common code into functions?

File: tools/crash-consistency-job.sh [hunk 8]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
```
Comment (chaitalisg) [inline tools/crash-consistency-job.sh:1]: Is this needed now that the test is moved under recovery_test package?

File: tests/robustness/filehandler/filehandler.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
+//go:build darwin || (linux && amd64)
```
Comment (chaitalisg) [inline tests/robustness/filehandler/filehandler.go:1]: This should be moved under `recovery` as well.

File: tests/recovery/recovery_test/main_test.go [hunk 10]
```diff
@@ -13,6 +13,8 @@ import (
 
 const (
 	dataSubPath = "recovery-data"
+	dirPath     = "kopia_dummy_repo"
```
Comment (chaitalisg) [inline tests/recovery/recovery_test/main_test.go:16]: Can you use a more descriptive name? Something like consistencyTestRepo or similar? Same thing goes for the path `kopia_dummy_repo`.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 11]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/kopia/kopia/repo/blob"
 	"github.com/kopia/kopia/snapshot"
 	"github.com/kopia/kopia/tests/robustness"
+	"github.com/kopia/kopia/tests/robustness/filehandler"
```
Comment (chaitalisg) [inline tests/recovery/blobmanipulator/blobmanipulator.go:19]: Filehandler is no longer under robustness package

File: tests/recovery/recovery_test/recovery_test.go [hunk 12]
```diff
@@ -168,6 +172,167 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestConsistencyWhenKill9AfterModify(t *testing.T) {
```
Comment (chaitalisg) [inline tests/recovery/recovery_test/recovery_test.go:175]: Add a comment describing the test function.

File: tests/recovery/recovery_test/recovery_test.go [hunk 13]
```diff
@@ -168,6 +172,167 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	err = bm.VerifySnapshot()
+	require.NoError(t, err)
+
+	// Create a temporary dir to restore a snapshot
+	restoreDir := t.TempDir()
+	if restoreDir == "" {
+		t.FailNow()
+	}
+
+	// try to restore a snapshot, this should error out
```
Comment (chaitalisg) [inline tests/recovery/recovery_test/recovery_test.go:253]: Comment says this should error out but the next line checks for `NoError`. Old comment?

File: Makefile [hunk 14]
```diff
@@ -307,6 +307,13 @@ recovery-tests: build-integration-test-binary $(gotestsum)
 	FIO_DOCKER_IMAGE=$(FIO_DOCKER_TAG) \
 	$(GO_TEST) -count=$(REPEAT_TEST) github.com/kopia/kopia/tests/recovery/recovery_test $(TEST_FLAGS)
 
+crash-consistency-test: export KOPIA_EXE ?= $(KOPIA_INTEGRATION_EXE)
```
Comment (chaitalisg) [inline Makefile:310]: Remove this if the target is no longer used

File: tests/recovery/recovery_test/main_test.go (hunk 10 above)
Comment (chaitalisg) [reply tests/recovery/recovery_test/main_test.go:16]: Please address this. The variable name `dirPath` does not describe its purpose. How about crashConsistencyDirPath or whatever the purpose of this variable is.
Similarly, `kopia_dummy_repo` does not describe its purpose. How about something like - crash-consistency-test-repo

**Comment by chaitalisg** [comment]
Please add a detailed PR description.

**Review by chaitalisg — APPROVED** [review]
LGTM!

File: tests/recovery/recovery_test/recovery_test.go [hunk 15]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+		t.FailNow()
+	}
+
+	bm, err := blobmanipulator.NewBlobManipulator(baseDir, dataRepoPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			t.Skip("Skipping crash consistency tests because KOPIA_EXE is not set")
+		} else {
+			t.Skip("Error creating SnapshotTester:", err)
+		}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:191]: You do not need the else clause because Skip() will exit the test.

```suggestion

		t.Skip("Error creating SnapshotTester:", err)
```

File: tests/recovery/recovery_test/recovery_test.go [hunk 16]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	}
+
+	bm.DataRepoPath = dataRepoPath
+
+	// create a snapshot for initialized data
+	snapID, err := bm.SetUpSystemWithOneSnapshot()
+	if err != nil {
+		t.FailNow()
+	}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:200]: How about using require package here too ? The advantage is that it'll actually also print the error which is good for debugging later.

```suggestion
	require.NoError(t, err)
```

File: tests/recovery/recovery_test/recovery_test.go [hunk 17]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+
+	cmpDir := bm.PathToTakeSnapshot
+
+	copyDir := t.TempDir()
+	if copyDir == "" {
+		t.FailNow()
+	}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:207]:
```suggestion
	require.NotEmpty(copyDir, "TempDir() did not generate a valid dir")
```

File: tests/recovery/recovery_test/recovery_test.go [hunk 18]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	err = bm.FileHandler.CopyAllFiles(bm.PathToTakeSnapshot, copyDir)
+	require.NoError(t, err)
+
+	// modify original files
+	content := "\nthis is a test for TestConsistencyWhenKill9AfterModify\n"
+	err = bm.FileHandler.ModifyDataSetWithContent(copyDir, content)
+	require.NoError(t, err)
+	log.Println("Copy content and modify the files.")
+
+	// kill the kopia command before it exits
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:228]: is this comment relevant here ?

You do not want to killl the `connect` command, but the snapshot command. Maybe move this comment to the snapshot command below ?

File: tests/recovery/recovery_test/recovery_test.go [hunk 19]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	log.Println("Verify snapshot corruption:")
+	// verify snapshot corruption
+	err = bm.VerifySnapshot()
+	require.NoError(t, err)
+
+	// Create a temporary dir to restore a snapshot
+	restoreDir := t.TempDir()
+	if restoreDir == "" {
+		t.FailNow()
+	}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:253]: same as before use the `require` package with a message explaining that TempDir() failed

File: tests/recovery/recovery_test/recovery_test.go [hunk 20]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = bm.FileHandler.CopyAllFiles(bm.PathToTakeSnapshot, copyDir)
+	require.NoError(t, err)
+
+	// modify original files
+	content := "\nthis is a test for TestConsistencyWhenKill9AfterModify\n"
+	err = bm.FileHandler.ModifyDataSetWithContent(copyDir, content)
+	require.NoError(t, err)
+	log.Println("Copy content and modify the files.")
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:226]: You should use `t.Log()` and `t.Logf()` where possible to avoid test noise.

SEE: https://pkg.go.dev/testing#B.Logf

These logs are generally not printed on the console, unless you specify the verbose `-v` flag to the test. Like `go test -v`.

The reason for using `t.Log` and `t.Logf` over `log.PrintX` is because developers usually run all the tests quite often. Every time we push anything to Kopia, we'd run all these tests locally. And if these tests would keep printing noise to the screen then it is hard to follow progress of all the tests. If anyone is interested in watching what a test is doing then they can specify the verbose flag or if the test fails then the test framework will dump the Log() statements anyway, so that they are there when you need them.

File: tests/recovery/recovery_test/recovery_test.go [hunk 21]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+	env := []string{"KOPIA_PASSWORD=" + testenv.TestRepoPassword}
+	cmd.Env = append(os.Environ(), env...)
+
+	o, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+	log.Println(string(o))
+
+	cmd = exec.Command(kopiaExe, "snap", "create", copyDir, "--json", "--parallel=1")
+
+	log.Println("Kill the kopia command before it exits:")
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:241]: Same suggestion as above, you can replace all your `log.PrintX` statements with `t.Log`.

File: tests/recovery/recovery_test/recovery_test.go [hunk 22]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+
+				break
+			}
+		}
+	}()
+
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		panic(err)
+	}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:308]: No need to panic() just fail the test. So, just use require.NoError().

File: tests/recovery/recovery_test/recovery_test.go (hunk 15 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: I deleted the else clause. Thanks~

File: tests/recovery/recovery_test/recovery_test.go (hunk 17 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go:206]: Thanks, I would like to use `require` interface instead.

File: tests/recovery/recovery_test/recovery_test.go (hunk 18 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go:227]: Yes, you are right. I do the kill action before connect, so I should add connect with repository with the environment configuration, otherwise it will display "ERROR open repository: repository is not connected.kopia connect repo".

File: tests/recovery/recovery_test/recovery_test.go (hunk 20 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go:225]: That's why I find it confusing when I see `t.Log` used in some places and `log.Println` used in others. I appreciate your response.

File: tests/recovery/recovery_test/recovery_test.go [hunk 23]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+			output := scanner.Text()
+			log.Println(output)
+			errOut.Write(scanner.Bytes())
+			errOut.WriteByte('\n')
+
+			log.Println(output)
+			// Check if the output contains the "hashing" etc.
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") || strings.Contains(output, "Snapshotting") {
+				log.Println("Detaching and terminating target process")
+				cmd.Process.Kill()
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:292]: This will work most of the times but will fail some of the times. The problem is that your strings.Contains() will not match every word every time this test runs.

`scanner := bufio.NewScanner(stderrPipe)` will allocate the default buffer size of 64k. What this means is that it MIGHT read upto those many characters before the first `scanner.Scan()` succeeds. In an unexpected scenario for your code, it might succeed once with half the word "uploa" first and the remaining "ded" in the next iteration, which will not make your test trigger the kill() action.

Which means that for your experiment, the kopia subcommand could actually print everything and then exit before the first iteration even triggering the kill action. Which is too late for you. Instead, I think what you are trying to do is to kill the subprocess as soon as you see lines being printed with the words you want.

So, you might be able to set a word split using `scanner.Split(bufio.ScanWords)` and check for words "uploaded" in that or you may also use `scanner.Split(bufio.ScanLines)` and then check for all the words in each line.

File: tests/recovery/recovery_test/recovery_test.go [hunk 24]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+
+	wg.Add(1)
+
+	o := bytes.Buffer{}
+
+	go func() {
+		defer wg.Done()
+
+		// Create a scanner to read from stdoutPipe
+		scanner := bufio.NewScanner(stdoutPipe)
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:310]: Do you really need to check BOTH stdout and stderr pipes ? I think you should be able to drop one of them.

File: tests/recovery/recovery_test/recovery_test.go (hunk 16 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: Fix it with `require.NoError(t, err)`

File: tests/recovery/recovery_test/recovery_test.go (hunk 17 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: fix it with `require.NotEmpty(copyDir, "TempDir() did not generate a valid dir")`

File: tests/recovery/recovery_test/recovery_test.go (hunk 19 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: fix it with require.NotEmpty(restoreDir, "TempDir() did not generate a valid dir")

File: tests/recovery/recovery_test/recovery_test.go (hunk 21 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: Fix it with `t.Log`

File: tests/recovery/recovery_test/recovery_test.go (hunk 22 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: fix it with `require.NoError()`

File: tests/recovery/recovery_test/recovery_test.go [hunk 25]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+			// Check if the output contains the "copying" text
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") {
+				cmd.Process.Kill()
+				break
+			}
+		}
+	}()
+
+	// Run the command
+	_ = cmd.Run()
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:328]: cmd.Run() will wait for the process to exit and it will also close the StdErr pipe. So, you should use use `cmd,Start()` instead and then wait for your go routine `wg.Wait()` and finally, wait for the command to exit using `cmd.Wait()`.

File: tests/recovery/recovery_test/recovery_test.go (hunk 24 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go:311]: I believe it's still necessary to have the stdoutpipe because there's a possibility that the snapshot will be created successfully. Although it's unlikely for the stdoutpipe to be executed in such cases, there are scenarios where the test environment is not cleaned properly in pipeline and the previous dataset is used for running the test. In such cases, it's important to know whether the snapshot was created successfully, even though the occurrence of this situation is unlikely.

File: tests/recovery/recovery_test/recovery_test.go [hunk 26]
```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+			output := scanner.Text()
+			log.Println(output)
+			o.Write(scanner.Bytes())
+			o.WriteByte('\n')
+
+			log.Println("snapshot create successfully", output)
+			// Check if the output contains the "copying" text
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") {
+				cmd.Process.Kill()
+				break
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:322]: You should drain the `stdoutPipe` before stopping the goroutine or your subcommand can get deadlocked. If a subprocess is blocked on writing to the std-err before of buffer limitations then you will not be able to kill it and hence your `cmd.Wait()` or `cmd.Run()` will get stuck forever.

File: tests/recovery/recovery_test/recovery_test.go (hunk 24 above)
Comment (Shrekster) [reply tests/recovery/recovery_test/recovery_test.go:311]: It should not matter whether if snapshot was created successfully. If Kopia prints to stderr then it will always print that line to stderr even if it succeeded.

File: tests/recovery/recovery_test/recovery_test.go (hunk 26 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go:296]: Since the log messages related to uploading are only printed to the stderrPipe, I believe there is no need to include the stdoutPipe anymore. Therefore, I think there is no requirement to drain the stdoutPipe in this scenario.

File: tests/recovery/recovery_test/recovery_test.go (hunk 24 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: I decide to delete stdoutPipe because the log about uploading only print at stderrPipe.

File: tests/recovery/recovery_test/recovery_test.go (hunk 25 above)
Comment (MinyiZou) [reply tests/recovery/recovery_test/recovery_test.go]: I used `cmd.Start()` rather than `cmd.Run()` in this case

File: tests/recovery/recovery_test/recovery_test.go [hunk 27]
```diff
@@ -168,6 +172,143 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+func killOnCondition(t *testing.T, cmd *exec.Cmd) {
+	t.Helper()
+
+	stderrPipe, err := cmd.StderrPipe()
+	require.NoError(t, err)
+
+	// execute kill -9 while receive ` | 1 hashing, 0 hashed (65.5 KB), 0 cached (0 B), uploaded 0 B, estimating...` message
+	var wg sync.WaitGroup
+
+	var mu sync.Mutex
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:269]: why do you need a mutex ?

File: tests/recovery/recovery_test/recovery_test.go [hunk 28]
```diff
@@ -168,6 +172,143 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
...
+
+	// Add a WaitGroup counter for the first goroutine
+	wg.Add(1)
+
+	errOut := bytes.Buffer{}
```
Comment (Shrekster) [inline tests/recovery/recovery_test/recovery_test.go:274]: is anyone using the errOut variable? If not then please drop it

File: tests/recovery/filehandler/filehandler.go [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+		sourceFile.Close()
+
+		destinationFile.Close()
+	}
+
+	return nil
+}
+
+// CompareDirs examines and compares the quantities and contents of files in two different folders.
+func (handler *FileHandler) CompareDirs(source, destination string) error {
```
Comment (Shrekster) [inline tests/recovery/filehandler/filehandler.go:115]: You can use the directory comparator fro the Kopia's internal `diff` module:

```go
	c, err := diff.NewComparer(&buf)
	require.NoError(t, err)

	t.Cleanup(func() {
		_ = c.Close()
	})

	err = c.Compare(ctx, dir1, dir2)
	require.NoError(t, err)
```

See this file: internal/diff/diff.go

File: tests/recovery/filehandler/filehandler.go [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+// FileHandler implements a FileHandler struct.
+type FileHandler struct{}
+
+// ErrFileNumberInconsistency is utilized to raise an error when the number of items in the two folders differs.
+var ErrFileNumberInconsistency = errors.New("the number of items in the two folders is not the same")
+
+// ErrFilesInconsistency is utilized to raise an error when the name of item in the two folders differs.
+var ErrFilesInconsistency = errors.New("the name of the item in the source folder does not match the name in the duplicated folder")
```
Comment (Shrekster) [inline tests/recovery/filehandler/filehandler.go:22]: you can drop this error now

File: tests/recovery/filehandler/filehandler.go [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+
```
Comment (Shrekster) [inline tests/recovery/filehandler/filehandler.go:23]: you can drop these errors now

```suggestion
```

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 32]
```diff
@@ -274,6 +276,54 @@ func (bm *BlobManipulator) SetUpSystemUnderTest() error {
...
+
+	// create snapshot of the data
+	log.Printf("Creating snapshot of directory %s", bm.PathToTakeSnapshot)
+
+	snapshotID, _, err := bm.TakeSnapshot(bm.PathToTakeSnapshot)
+	if err != nil {
+		return "", err
+	}
+
+	log.Printf("Creating snapshot of snapshot ID %s", snapshotID)
```
Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:302]:
```suggestion
	log.Printf("Created snapshot of snapshot ID %s", snapshotID)
```

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 33]
```diff
@@ -274,6 +276,54 @@ func (bm *BlobManipulator) SetUpSystemUnderTest() error {
...
+	numFiles := 50
+	ctx := context.Background()
+
+	err := bm.writeRandomFiles(ctx, fileSize, numFiles)
+	if err != nil {
+		return "", err
+	}
+
+	// create snapshot of the data
+	bm.PathToTakeSnapshot = bm.FileHandler.GetRootDir(bm.fileWriter.DataDirectory(ctx))
```
Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:292]: why not set the `bm.PathToTakeSnapshot` to the data dir ? We'd like to take the snapshot of the entire directory, right ?

```suggestion
	bm.PathToTakeSnapshot = bm.fileWriter.DataDirectory(ctx)
```

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 34]
```diff
@@ -274,6 +276,54 @@ func (bm *BlobManipulator) SetUpSystemUnderTest() error {
...
+// Tests can later create snapshots from the source directory.
+func (bm *BlobManipulator) GenerateRandomFiles(fileSize, numFiles int) error {
+	ctx := context.Background()
+
+	err := bm.writeRandomFiles(ctx, fileSize, numFiles)
+	if err != nil {
+		return err
+	}
+
+	bm.PathToTakeSnapshot = bm.FileHandler.GetRootDir(bm.fileWriter.DataDirectory(ctx))
```
Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:317]: same suggestions as above here:

```suggestion
	bm.PathToTakeSnapshot = bm.fileWriter.DataDirectory(ctx)
```

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 33 above)
Comment (Shrekster) [reply tests/recovery/blobmanipulator/blobmanipulator.go:292]: Maybe I am missing something here. @MinyiZou I am assuming hat we'll ike to snapshot the entire directory and not just a single file.

Comment (MinyiZou) [reply tests/recovery/blobmanipulator/blobmanipulator.go:292]: The structure of directories is as follows: /a/b/. In the "b" folder, there are 200 files. However, when using the bm.fileWriter.DataDirectory(ctx) method, I only receive a folder path (/a/). For this particular test scenario, it should be acceptable to capture either the entire directory as a snapshot or the 200 individual files. In case we choose to snapshot the 200 files, I will send a signal. `bm.FileHandler.GetRootDir` will return /a/b/ path if I input /a/.

**Review by Shrekster — APPROVED** [review]
LGTM! Nice work!

A couple things around the raciness attribute of the current design of the crash test, but this a good first iteration.
